src/WCSimIBDGen.cc

```diff
@@ -0,0 +1,258 @@
+        rand_cos_theta = -1.0 + 2.0 * G4UniformRand();
+
+        // Weight events by spectrum
+        G4double xs_test = xs_max * flux_max * G4UniformRand();
+
+        // Cross section
+        G4double xs_weight = CrossSection(rand_ene, rand_cos_theta);
+
+        // Flux at rand_ene
+        G4double flux_weight = InterpolateSpectrum(energy, flux, rand_ene);
+
+        passed = (xs_test < xs_weight * flux_weight);
+    }
+}
+
+double WCSimIBDGen::CrossSection(double e_nu, double cos_theta) {
+
+    const double cos_theta_c = (0.9741 + 0.9756) / 2.0;
+
+    // Radiative correction constant
+    const double rad_cor = 0.024;
+
+    const double DELTA = CLHEP::neutron_mass_c2 - CLHEP::proton_mass_c2;
+
+    // Neutrino energy threshold for inverse beta decay
+    const double e_nu_min = ((CLHEP::proton_mass_c2 + DELTA + CLHEP::electron_mass_c2) *
+                                 (CLHEP::proton_mass_c2 + CLHEP::electron_mass_c2 + DELTA) -
+                             CLHEP::proton_mass_c2 * CLHEP::proton_mass_c2) /
+                            2 / CLHEP::proton_mass_c2;
+
+    if (e_nu < e_nu_min) {
+        return 0.0;
+    }
+
+    const double GFERMI = 1.16639e-11;
+
+    const double sigma_0 = GFERMI * GFERMI * cos_theta_c * cos_theta_c / CLHEP::pi * (1 + rad_cor);
+
+    // Couplings
+    const double f = 1.00;
+    const double f2 = 3.706;
+    const double g = 1.26;
+
+    // Order 0 terms
+    double e0 = e_nu - DELTA;
+    if (e0 < CLHEP::electron_mass_c2) {
+        e0 = CLHEP::electron_mass_c2;
+    }
+    double p0 = sqrt(e0 * e0 - CLHEP::electron_mass_c2 * CLHEP::electron_mass_c2);
+    double v0 = p0 / e0;
+
+    // Order 1 terms
+    const double y_squared = (DELTA * DELTA - CLHEP::electron_mass_c2 * CLHEP::electron_mass_c2) / 2;
+    double e1 = e0 * (1 - e_nu / CLHEP::proton_mass_c2 * (1 - v0 * cos_theta)) - y_squared / CLHEP::proton_mass_c2;
+    if (e1 < CLHEP::electron_mass_c2) {
+        e1 = CLHEP::electron_mass_c2;
+    }
+    double p1 = sqrt(e1 * e1 - CLHEP::electron_mass_c2 * CLHEP::electron_mass_c2);
+    double v1 = p1 / e1;
+
+    double gamma =
+        2 * (f + f2) * g *
+            ((2 * e0 + DELTA) * (1 - v0 * cos_theta) - CLHEP::electron_mass_c2 * CLHEP::electron_mass_c2 / e0) +
+        (f * f + g * g) * (DELTA * (1 + v0 * cos_theta) + CLHEP::electron_mass_c2 * CLHEP::electron_mass_c2 / e0) +
+        (f * f + 3 * g * g) * ((e0 + DELTA) * (1 - cos_theta / v0) - DELTA) +
+        (f * f - g * g) * ((e0 + DELTA) * (1 - cos_theta / v0) - DELTA) * v0 * cos_theta;
+
+    double cross_section = ((f * f + 3 * g * g) + (f * f - g * g) * v1 * cos_theta) * e1 * p1 -
+                           gamma / CLHEP::proton_mass_c2 * e0 * p0;
+
+    cross_section *= sigma_0 / 2;
+
+    // Convert from MeV^{-2} to mm^2 (native units for GEANT4)
+    cross_section *= CLHEP::hbarc * CLHEP::hbarc;
+
+    return cross_section;
+}
+
+double WCSimIBDGen::PositronEnergy(double e_nu, double cos_theta) {
```

Comment:
Can you add some comments to say which paper(s) this correction came from?

Reply:
Comment added 👍